neqo-transport/src/connection.rs

```diff
@@ -2568,6 +2582,46 @@ mod tests {
         assert!(server.crypto.tls.info().unwrap().resumed());
     }
 
+    fn connect_with_rtt(client: &mut Connection, server: &mut Connection, now: &mut Instant) {
+        const TIME_INTERVAL: u64 = 20;
+
+        let mut a = client;
+        let mut b = server;
+        let mut datagram = None;
+        let is_done = |c: &mut Connection| match c.state() {
+            State::Confirmed | State::Closing { .. } | State::Closed(..) => true,
+            _ => false,
+        };
+
+        while !is_done(a) {
+            let _ = maybe_authenticate(a);
+            let d = a.process(datagram, *now);
+            datagram = d.dgram();
+            *now += Duration::from_millis(TIME_INTERVAL);
+            mem::swap(&mut a, &mut b);
+        }
+        a.process(datagram, *now);
+    }
+
+    #[test]
+    fn remember_smoothed_rtt() {
+        let mut client = default_client();
+        let mut server = default_server();
+
+        let mut now_time = now();
+        connect_with_rtt(&mut client, &mut server, &mut now_time);
+
+        // In exchange_ticket we only send a packet from a server to a client
+        // and client won't calculate rtt from that packet, so no need to increase time here.
+        let token = exchange_ticket(&mut client, &mut server);
+        let mut client = default_client();
+
+        client
+            .set_resumption_token(now(), &token[..])
+            .expect("should set token");
+        assert_eq!(client.loss_recovery.latest_rtt(), Duration::from_millis(40));
+    }
```

Comment:
Once you have tested that the RTT is set when you set a resumption token, I would then test to see if that estimate affects the RTT estimate if the RTT changes.  For instance, you have a 40ms RTT simulated here.  If you make a new connection with a 50ms RTT, then it should report 50ms.  If you incorrectly added the 40ms value from the previous connection as a sample, then the rolling average will include that and you will likely get something like 48 or 49ms instead.

Reply:
@martinthomson seems either I messed everything up while doing these changes...or I extending the test in a wrong way. What I've just tried to do is straight after these lines add smth like

```rust 
        // Add one more client to connect it with the new rtt
        let NEW_RTT  = Duration::from_millis(90);
        let mut client = default_client();
        connect_with_rtt(&mut client, &mut server, &mut now_time, NEW_RTT);
        let token = exchange_ticket(&mut client, &mut server);
        // Again add one more client to check its rtt
       let mut client = default_client();
        client
            .set_resumption_token(now(), &token[..])
            .expect("should set token");
        assert_eq!(client.loss_recovery.rtt(), NEW_RTT);
```

Suspect my understanding how to do it is wrong. Could you plz give me a hand with this? Maybe I've also broken smth with my last commit)

Reply:
I think that what you want is something like (excuse lack of formatting):

```rust
const RTT1: Duration = Duration::from_millis(50);
const RTT2: Duration = Duration::from_millis(80);
let mut now = now();
let mut client = default_client();
now = connect_with_rtt(&mut client, &mut server, now, RTT1);
assert_eq!(client.loss_recovery.rtt(), RTT1);
let token = exchange_ticket(&mut client, &mut server);

let mut client = default_client();
client.set_resumption_token(now, &token[..]).expect("should set token");
assert_eq!(client.loss_recovery.rtt(), RTT1, "client should remember previous RTT");

now = connect_with_rtt(&mut client, &mut server, now, RTT2);
assert_eq!(client.loss_recovery.rtt(), RTT2, "previous RTT should be completely erased");
```

Reply:
Thanks. The last assert fails. So looks like it does not forget the previous RTT.

Reply:
Hmm, that's unexpected.  Can you push the code, maybe with that test commented out, and I'll take a look?

Reply:
@martinthomson I commented out the failing part of the test. It would be great if you could take a look at it.